File: lang_lawyer/python/div_mod_sign.py
print('(n, (n // 3, n % 3), (n // -3, n % -3))')

for i in range(-6, 7):
    print((i, divmod(i, 3), divmod(i, -3)))

"""
Results:
(floor division - towards -inf, modulus = dividend - result * divisor, e.g.:
2 % (-3) == 2 - (-1) * (-3) == 2 - 3 == -1)

This means that the modulus has the sign of the DIVISOR.

but: (!!!)

modulus != sign(divisor) * (abs(dividend) % abs(divisor))

(-6, (-2, 0), (2, 0))
(-5, (-2, 1), (1, -2))
(-4, (-2, 2), (1, -1))
(-3, (-1, 0), (1, 0))
(-2, (-1, 1), (0, -2))
(-1, (-1, 2), (0, -1))
(0, (0, 0), (0, 0))
(1, (0, 1), (-1, -2))
(2, (0, 2), (-1, -1))
(3, (1, 0), (-1, 0))
(4, (1, 1), (-2, -2))
(5, (1, 2), (-2, -1))
(6, (2, 0), (-2, 0))
"""

File: lang_lawyer/c++/div_mod_sign.cpp
#include <iostream>

using namespace std;

int main()
{
    cout << "(n, (n // 3, n % 3), (n // -3, n % -3))" << endl;
    for (int i=-6; i < 7; ++i)
    {
        cout << '(' << i << ", (" << i / 3 << ", " << i % 3 << "), "
             << '(' << i / -3 << ", " << i % -3 << "))" << endl;
    }

    return 0;
}

/* Results:
(trunc division - towards 0, modulus == dividend - divisor * result, e.g.:
2 % (-3) == 2 - (-3) * 0 == 2)

This means that the  modulus has the sign of the DIVIDEND

and:

modulus == sign(dividend) * (abs(dividend) % abs(divisor))

(-6, (-2, 0), (2, 0))
(-5, (-1, -2), (1, -2))
(-4, (-1, -1), (1, -1))
(-3, (-1, 0), (1, 0))
(-2, (0, -2), (0, -2))
(-1, (0, -1), (0, -1))
(0, (0, 0), (0, 0))
(1, (0, 1), (0, 1))
(2, (0, 2), (0, 2))
(3, (1, 0), (-1, 0))
(4, (1, 1), (-1, 1))
(5, (1, 2), (-1, 2))
(6, (2, 0), (-2, 0))

*/

